### ISSUE TITLE
Feature Proposal: 'max-len' and 'max-lines' options for arrow-body-style 'as-needed'

### ISSUE THREAD

**Issue by low-ghost** [eslint/eslint#5504]
**Version**
2.3.0

**Feature**
Arrow functions exceeding a certain character length or line length might deserve brackets. This feature allows adding options 'max-len' and 'max-lines' that will apply to arrow-body-style with 'as-needed'. These are considered problems:
```js
/*eslint arrow-body-style: [2,  "as-needed", { "max-len": 30 }]*/

var foo = () => Promise.resolve(bar()).then(bar).then(bar);
```
```js
/*eslint arrow-body-style: [2,  "as-needed", { "max-lines": 2 }]*/

var foo = () => 
    Promise.resolve(bar())
        .then(bar)
        .then(bar);
```
The following are not:
```js
/*eslint arrow-body-style: [2,  "as-needed", { "max-len": 30 }]*/

var foo = () => {
    return Promise.resolve(bar()).then(bar).then(bar);
};
```
```js
/*eslint arrow-body-style: [2,  "as-needed", { "max-lines": 2 }]*/

var foo = () => {
    return Promise.resolve(bar())
        .then(bar)
        .then(bar);
};
```
**Solution**
Pretty straightforward. See PR.

**Comment by low-ghost** [comment]
also not a problem:
```js
/*eslint arrow-body-style: [2,  "as-needed", { "max-len": 30 }]*/

var foo = () => bar();
```

**Comment by ilyavolodin** [comment]
In general sounds like a good idea, but I don't use arrow function. @eslint/eslint-team what do you think?
P.S. Sorry, lost track of this issue, so it never got triage.

**Comment by gyandeeps** [comment]
Just to flush out this bit more, what happens in the following cases:

```js
// case 1
var foo = () => function(){
	var a = 1;
  	
  	a = 2;
  	a = 3;
  	a = 4;
}

//case 2
var foo = () => 



    Promise.resolve(bar())
        .then(bar)
        .then(bar);

// case 3
var foo = () => 
    Promise.resolve(bar())

	.then(bar)
    

	.then(bar);


```

**Comment by low-ghost** [comment]
Good thoughts, I'll add a few more tests in the morning. For 2 and 3,  the declaration and each line, including whitespace, count as a line. So glancing on a phone, both are about 9 lines while the whitespace doesn't effect character count. Desired effect? I played around with the declaration not being included in the count, so 
```js
var foo = () =>
    bar();
```
being technically one line, but ended up siding against it because it isn't quite intuitive.

Case 1 is a 7 line body, but the function could be a little tricky, I'm not sure exactly how it'd show up in the AST. Presumably would thow an error, but I don't believe it should. Especially in cases with nested arrows:

```js
var foo = () => () =>
    bar();
```
with a max-lines of 1 should only throw for the innermost function such that

```js
var foo () => () => {
    return bar();
};
```
would pass. I'll definitely get on checking that.

**Comment by nzakas** [comment]
I'm :-1: on this suggestion. I think maximums like this are a bit arbitrary and confusing. I'd suggest creating a custom rule for these options, as there's no direct dependency what arrow-body-style does.

**Comment by captbaritone** [comment]
I'd like to echo @nzakas' skepticism of length/size based rules. CSSlint uses this style of rule extensively (eg no more than 10 floats per file) and it has been a huge failure in our code base. The person who merges the change that breaks the proverbial camel's back ends up being unfairly responsible for fixing an endemic problem. The result is either skipping Git hooks all together, or awkwardly rewriting code simply to avoid triggering the linter, since that is generally simpler than refactoring the surrounding code.

Obviously this specific suggestion is less susceptible to a failure mode like this, but I would be remiss if I didn't warn against the possible dangers of this class of issue.

**Comment by low-ghost** [comment]
@nzakas and @captbaritone, I can see that. I was thinking it fit somewhere within the realm of rules like "max-params", "max-length", "max-statements", "complexity" and so on, rules intentionally attempting to limit potential complexity by establishing maximums. It could be misplaced though. Perhaps this would be better solved by combinations of these existing rules; "max-params" and "max-depth" alone might go a decent part of the way, maybe something like a "max-chained". However, the codebase I'm working on definitely points to some necessity for a limit. It's so damn exciting to skip the block and return that suddenly you're 100 lines into an incredibly confusing nest of arrow functions and promises. And yet I don't want to give up the power of short lambdas like `map(() => doSomething(bar))` by forcing "always". Happy to just keep it as a custom rule though.

**Comment by low-ghost** [comment]
@gyandeeps added tests for cases 2 and 3 showing that they work as expected and some added functionality for case 1. With these changes, case 1 passes with a max-lines of 1. This is handled by only validating from the topmost arrow to the start of the next nested function or arrow expression. This means that this passes with max-lines 1:
```js
var foo () => () => {
    var a = 1;
    return bar(a);
};
```
while this does not:
```js
var foo () =>
    () => {
        var a = 1;
        return bar(a);
    };
```
because the first arrow function starts at the line 1 at the first `=>` and terminates on the next line at `()`. Similarly, with `"include-args": true` and `"max-len": 10`, this fails:
```js
var foo (incrediblyLong, functionSignature, meansThis, shouldHave, aBlock, statementAfter) => () => {
   var a = 1;
    return bar(1);
};
```
Finally, enforcement of unexpected block warnings are not implemented for nested arrows. So this also is fine, even though it might seem like it should warn that bar doesn't need its own block:
```js
var foo () => () => {
    return bar();
};
```
This because there may be other reasons to provide a block here (such as a long line, to highlight the innermost functionality etc).

Still happy to keep this as a custom rule. Thanks for pointing out some edge cases.

**Comment by nzakas** [comment]
Yeah, I think this is best as a custom rule. This is part of why we make it easy to create custom rules: this is clearly something that you feel is useful but doesn't really fit in the core. 

Here are some docs to get you started:
http://eslint.org/docs/developer-guide/working-with-rules
http://eslint.org/docs/developer-guide/working-with-plugins
